File: src/main/java/net/tecdroid/subsystems/elevatorjoint/ElevatorJointConfiguration.java
package net.tecdroid.subsystems.elevatorjoint;

import edu.wpi.first.units.measure.Angle;
import edu.wpi.first.units.measure.Current;

import edu.wpi.first.units.measure.Time;
import net.tecdroid.subsystems.elevatorjoint.ElevatorJoint.*;
import net.tecdroid.util.*;

import static edu.wpi.first.units.Units.*;


public class ElevatorJointConfiguration {
    private static class DeviceIDS {
        private final DigitId mElevatorAngleDigit = new DigitId(6);
        private final DigitId mLeadingMotorDigit = new DigitId(1);
        private final DigitId mFollowerMotorDigit = new DigitId(2);
        private final DigitId TroughBorePort = new DigitId(3);

        final DeviceIdentifiers elevatorAngleDevicesIDs = new DeviceIdentifiers(
                IdentifiersKt.joinDigits(mElevatorAngleDigit, mLeadingMotorDigit),
                IdentifiersKt.joinDigits(mElevatorAngleDigit, mFollowerMotorDigit),
                TroughBorePort
        );
    }

    private static class Devices {
        private final MotorProperties elevatorAngleMotorProperties = Motors.INSTANCE.getKrakenX60();

        final DeviceProperties elevatorAngleDeviceProperties = new DeviceProperties(
                elevatorAngleMotorProperties
        );
    }

    private static class Limits {
        private final Current motorsCurrent = Amps.of(40);
        private final Angle minimumElevatorAngle = Degrees.of(10.0);
        private final Angle maximumElevatorAngle = Degrees.of(90.0);
        private final Angle gravityPointAngle = Degrees.of(75.0);

        final DeviceLimits elevatorAngleDeviceLimits = new DeviceLimits (
                motorsCurrent, minimumElevatorAngle, maximumElevatorAngle,
                gravityPointAngle
        );
    }

    private static class Conventions {
        private final RotationalDirection mElevatorAngleMotorsPositiveDirection = RotationalDirection.Clockwise;

        final DeviceConventions elevatorAngleDeviceConventions = new DeviceConventions(
                mElevatorAngleMotorsPositiveDirection
        );

    }

    private static class Structure {
        private final GearRatio elevatorAngleMotorGR = new GearRatio(360, 1, 0);
        private final Angle encoderOffset = Degrees.of(0.0);
        final PhysicalDescription elevatorAnglePhysicalDescription = new PhysicalDescription(
                elevatorAngleMotorGR, encoderOffset
        );
    }

    private static class Control {
        private final PidfCoefficients elevatorAnglePidfCoefficients = new PidfCoefficients(0.0, 0.0, 0.0, 0.0);

        // Svag gains when we have no gravity Slot0
        private final SvagGains elevatorAngleNoGravitySvagGains = new SvagGains(0.0, 0.0, 0.01, 0.0);

        // Svag gains when we have gravity Slot1
        private final SvagGains elevatorAngleGravitySvagGains = new SvagGains(0.0, 0.0, 0.01, 0.0);

        private final MotionMagicSettings elevatorAngleMotionMagicCoefficients = new MotionMagicSettings(0.0, 0.0, 0.0);
        private final Time elevatorAngleRampRate = Seconds.of(0.1);

        final ControlConstants elevatorAngleControlConstants = new ControlConstants(
                elevatorAnglePidfCoefficients, elevatorAngleNoGravitySvagGains,
                elevatorAngleGravitySvagGains, elevatorAngleMotionMagicCoefficients,
                elevatorAngleRampRate
        );

    }

    public static final Config elevatorJointConfig = new Config(
            new DeviceIDS().elevatorAngleDevicesIDs, new Devices().elevatorAngleDeviceProperties,
            new Limits().elevatorAngleDeviceLimits, new Conventions().elevatorAngleDeviceConventions,
            new Structure().elevatorAnglePhysicalDescription, new Control().elevatorAngleControlConstants
    );
}

File: src/main/java/net/tecdroid/subsystems/elevatorjoint/ElevatorJointConfiguration.kt
package net.tecdroid.subsystems.elevatorjoint

import edu.wpi.first.units.AngleUnit
import edu.wpi.first.units.Units.Second
import edu.wpi.first.units.measure.Angle
import edu.wpi.first.units.measure.Current
import net.tecdroid.mechanical.Reduction
import net.tecdroid.safety.MeasureLimits
import net.tecdroid.util.*
import net.tecdroid.util.RotationalDirection.Counterclockwise
import net.tecdroid.util.amps
import net.tecdroid.util.rotations
import net.tecdroid.util.seconds

data class ElevatorJointConfig(
    val leadMotorControllerId: NumericId,
    val followerMotorControllerId: NumericId,
    val absoluteEncoderPort: NumericId,
    val absoluteEncoderIsInverted: Boolean,
    val motorDirection: RotationalDirection,
    val motorCurrentLimit: Current,
    val reduction: Reduction,
    val measureLimits: MeasureLimits<AngleUnit>,
    val controlGains: ControlGains,
    val motionTargets: AngularMotionTargets,
    val absoluteEncoderOffset: Angle
)

val elevatorJointConfig = ElevatorJointConfig(
    leadMotorControllerId = NumericId(51),
    followerMotorControllerId = NumericId(52),
    motorDirection = Counterclockwise,
    motorCurrentLimit = 40.0.amps,

    absoluteEncoderPort = NumericId(0),
    absoluteEncoderIsInverted = true,
    absoluteEncoderOffset = 0.4840.rotations,

    reduction = Reduction(360.0),

    measureLimits = MeasureLimits(
        absoluteMinimum = 0.011.rotations,
        relativeMinimum = 0.025.rotations,
        relativeMaximum = 0.26.rotations,
        absoluteMaximum = 0.2682.rotations,
    ),

    controlGains = ControlGains(
        p = 0.8,
        s = 0.16263, // 0.14604,
        v = 0.11085, // 0.11017,
        a = 0.002245, // 0.0035221,
        g = 0.01139, // 0.011399
    ),

    motionTargets = AngularMotionTargets(
        cruiseVelocity = 0.2.rotations.per(Second   ),
        accelerationTimePeriod = 0.25.seconds,
        jerkTimePeriod = 0.1.seconds
    )
)
